### ISSUE TITLE
Consider converting to TypeScript

### ISSUE THREAD

**Issue by jacobq** [mljs/levenberg-marquardt#23]
Personally, I have found writing in [TypeScript](https://www.typescriptlang.org/) to be very helpful in catching subtle mistakes (e.g. it would have caught the argument type mismatch in https://github.com/mljs/levenberg-marquardt/issues/14). It also has some perks like making auto This project is small so the conversion would not be difficult (and I'd be happy to do it), but I'd like to know the rest of the team's thoughts on this.

* Are you open to considering the idea of changing this project's language to TypeScript? If not, why not?
* If so, how do you like the idea of using [`rollup-plugin-typescript`](https://github.com/rollup/rollup-plugin-typescript)?

**Comment by targos** [comment]
Yes (but after merging https://github.com/mljs/levenberg-marquardt/pull/19)!

We started using TypeScript not long ago. See https://github.com/mljs?utf8=%E2%9C%93&q=&type=&language=typescript

We would not need rollup anymore. The 4 projects are based on this generator: https://github.com/cheminfo/generator-cheminfo/tree/master/generators/typescript

**Comment by targos** [comment]
BTW, the only dependency of this project (ml-matrix) recently got TypeScript definitions: https://github.com/mljs/matrix/blob/master/matrix.d.ts

**Comment by maasencioh** [comment]
This could be a good chance to learn some ts for me, do you mind if I do this? @jacobq @targos

**Comment by jacobq** [comment]
> This could be a good chance to learn some ts for me, do you mind if I do this?

No objections from me. I also don't see any reason why it would have to be done by just one person. My recommendation would be (after that other PR lands): create a branch (say `convert-to-typescript`), start making PRs to it, and get feedback via comments/reviews. Once everything is switched over and working then merge to master.
